contracts/tasks/smokeTest.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * Execute smoke tests on the mainnet:fork network.
+ *
+ * This script reads all of the tests present in the `contracts/scripts/test/smokeTest` folder. Each
+ * file should contain one tests and export these functions: 
+ *  - beforeDeploy (hre) -> function executes before contract deployment and can return validation data
+ *  - afterDeploy (hre, beforeDeployData) -> function executes after contract deployment and as a second
+ *    param receives return value of `beforeDeploy` function. `beforeDeployData` can be used to validate
+ *    that the state change is ok
+ */
+const { getDeployScripts, getLastDeployScript, getFilesInFolder } = require('../utils/fileSystem');
+
+async function getAllTests() {
+  const testsFolder = `${__dirname}/../scripts/test/smokeTest`;
+  const testFiles = await getFilesInFolder(testsFolder);
+  const tests = [];
+
+  testFiles.forEach(testFile => {
+    const test = require(`${testsFolder}/${testFile}`);
+    if (!test.beforeDeploy || test.beforeDeploy.length !== 1) {
+      throw new Error("Missing beforeDeploy(hre) smoke test function")
+    }
+    if (!test.afterDeploy || test.afterDeploy.length !== 2) {
+      throw new Error("Missing afterDeploy(hre, beforeDeployData) smoke test function")
+    }
+    test.name = testFile
+    tests.push(test)
+  })
+
+  return tests
+}
+
+async function smokeTest(taskArguments, hre) {
+  const deployId = taskArguments.deployid
+  let deployScript
+
+  if (deployId) {
+    const scripts = await getDeployScripts()
+    deployScript = scripts[parseInt(deployId)]
+  } else {
+    deployScript = await getLastDeployScript()
```

Comment:
It would be useful to be able to run smoke tests without doing a deploy.

For example, when I send a governance proposal to change a setting on the vault or a strategy, I'd like to be able to verify that it does not break anything on the fork.

So I would suggest that if no deployId is passed as an option on the command line, then we don't run any deploy. This way, I could:
 - start the fork
 - execute a governance proposal on the form manually (by running side script)
 - run the smoke test without passing a deployId

Reply:
If we are changing settings on the vault on the mainnet, wouldn't it make sense that we include all those governance / contract proposals into the deploy scripts? This way dev environment should more closely match the production one.

Besides the above point, would it be ok for you if when there is no `deployId` the node just runs in the background as it is running now. And the shell is waiting for user input to run smoke tests. So the usage should be: 
- run the smokeTest script without `deployid` param
- once node is ready use `hardhat console` command to connect to the node and do necessary changes
- press enter in the smokeTest terminal and all smoke tests are run. Once finished the process waits for user input to run them again. Exiting process (ctrl + C) kills the node and smoke tests.

This way we should have the versatility we require and also have the tests in a single script to make them CI ready in the future.

Reply:
Agreed that ideally there would be a migration for every governance proposal. But I think practically speaking it won't happen for a couple reasons: a) sometime it is urgent to flip a setting and also b) we would end up with dozens of migrations which would make unit tests setup really heavy. Practically speaking, what we have been doing is we try to keep the initial deploy script 001_core.js fairly up to date to match the settings on prod.

Your proposal sounds good. Let's try that. Thanks! :)

Reply:
Then again maybe there should be an additional stopping when all the tests execute the `before` step in the interactive mode.